File: sdks/java/fm-sdk/src/test/java/fm/internal/BehaviourFixturesTest.java
package fm.internal;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Stream;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import tools.jackson.databind.JsonNode;

import fm.Market;
import fm.Order;
import fm.OrderBook;
import fm.Side;
import fm.Trade;
import fm.Trades;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

/**
 * Every behaviour fixture, driven through this SDK's aggregators.
 *
 * <p>{@link WireFixturesTest} next door compares <em>parsed field values</em>:
 * one payload in, one set of fields out. It says nothing about what
 * {@link OrderBook} and {@link Trades} do with a sequence of them, which is
 * where the three SDKs have actually been wrong together — a book that
 * double-counts a cancel and a tape that holds its trades backwards both parse
 * every field correctly.
 *
 * <p>So these are inputs and answers rather than payloads and fields: a market,
 * a list of update steps, and what the aggregator must hold at the end. Java,
 * Python and TypeScript each run all of them, so a behaviour cannot be right in
 * one SDK and wrong in another without saying so.
 *
 * <p>See {@code sdks/fixtures/README.md}.
 */
class BehaviourFixturesTest {

    private static final Path FIXTURES =
            Path.of("..", "..", "fixtures", "behaviour").toAbsolutePath().normalize();

    private static final List<String> AGGREGATORS = List.of("OrderBook", "Trades");

    private record Fixture(String name, JsonNode document) {
        @Override
        public String toString() { return name; }
    }

    static Stream<Fixture> fixtures() throws Exception {
        try (Stream<Path> files = Files.list(FIXTURES)) {
            List<Fixture> found = new ArrayList<>();
            for (Path path : files.filter(p -> p.toString().endsWith(".json")).sorted().toList()) {
                found.add(new Fixture(
                        path.getFileName().toString().replace(".json", ""),
                        HttpFlexemarkets.MAPPER.readTree(Files.readString(path))));
            }
            return found.stream();
        }
    }

    private static Market market(JsonNode doc) {
        return HttpFlexemarkets.MAPPER.treeToValue(doc.get("market"), Market.class);
    }

    private static Order[] orders(JsonNode step) {
        List<Order> parsed = new ArrayList<>();
        for (JsonNode node : step.get("orders")) {
            parsed.add(HttpFlexemarkets.MAPPER.treeToValue(node, Order.class));
        }
        return parsed.toArray(new Order[0]);
    }

    private static List<List<Long>> levels(Map<Long, Long> book) {
        List<List<Long>> flattened = new ArrayList<>();
        book.forEach((price, units) -> flattened.add(List.of(price, units)));
        return flattened;
    }

    private static List<List<Long>> expectedLevels(JsonNode node) {
        List<List<Long>> wanted = new ArrayList<>();
        for (JsonNode level : node) {
            wanted.add(List.of(level.get(0).asLong(), level.get(1).asLong()));
        }
        return wanted;
    }

    @ParameterizedTest(name = "behaviour: {0}")
    @MethodSource("fixtures")
    void behaviour(Fixture fixture) {
        JsonNode doc = fixture.document();

        List<Long> delivered = new ArrayList<>();
        for (JsonNode step : doc.get("steps")) {
            for (JsonNode order : step.get("orders")) {
                delivered.add(order.get("id").asLong());
            }
        }
        List<Long> declared = new ArrayList<>();
        doc.get("deliveredIds").forEach(id -> declared.add(id.asLong()));
        assertEquals(declared, delivered,
                "the fixture's input is not in the order it declares. deliveredIds is "
                + "what stops a case being quietly reordered into one that proves "
                + "nothing — the trades-ordering fixture is only a test at all "
                + "because its input arrives newest first.");

        switch (doc.get("type").asString()) {
            case "OrderBook" -> checkOrderBook(doc);
            case "Trades" -> checkTrades(doc);
            default -> fail("no aggregator for type " + doc.get("type").asString());
        }
    }

    private void checkOrderBook(JsonNode doc) {
        OrderBook book = new OrderBook(market(doc));
        for (JsonNode step : doc.get("steps")) {
            if (step.path("clear").asBoolean(false)) {
                book.clear();
            }
            book.update(orders(step));
        }


        JsonNode expect = doc.get("expect");
        expect.propertyNames().forEach(key -> {
            JsonNode want = expect.get(key);
            switch (key) {
                case "bestBuyPrice" -> assertEquals(want.asLong(), book.bestBuyPrice(), key);
                case "bestBuyUnits" -> assertEquals(want.asLong(), book.bestBuyUnits(), key);
                case "bestSellPrice" -> assertEquals(want.asLong(), book.bestSellPrice(), key);
                case "bestSellUnits" -> assertEquals(want.asLong(), book.bestSellUnits(), key);
                case "hasValueBuy" -> assertEquals(want.asBoolean(), book.hasValue(Side.BUY), key);
                case "hasValueSell" -> assertEquals(want.asBoolean(), book.hasValue(Side.SELL), key);
                case "buyLevels" -> assertEquals(expectedLevels(want), levels(book.buyLevels()), key);
                case "sellLevels" -> assertEquals(expectedLevels(want), levels(book.sellLevels()), key);
                default -> fail("fixture asks for unknown key " + key);
            }
        });
    }

    private void checkTrades(JsonNode doc) {
        Trades tape = new Trades(market(doc));
        int index = 0;
        for (JsonNode step : doc.get("steps")) {
            if (step.path("clear").asBoolean(false)) {
                tape.clear();
            }

            Trade[] added = tape.update(orders(step));

            // What update() reports it added is what MarketView dispatches
            // onTrade from. A step that declares `adds` pins it -- including
            // the zero, which is the update a handler must stay silent through.
            if (step.has("adds")) {
                assertEquals(step.get("adds").asInt(), added.length,
                        "step " + index + " (" + step.path("note").asString("") + ") "
                        + "reported " + added.length + " new trades");
            }
            index++;
        }

        JsonNode expect = doc.get("expect");
        Trade[] held = tape.mostRecentTrades();

        if (expect.has("size")) {
            assertEquals(expect.get("size").asInt(), tape.size(), "size");
        }

        if (expect.has("trades")) {
            JsonNode wanted = expect.get("trades");
            assertEquals(wanted.size(), held.length,
                    "tape holds " + held.length + " trades, expected " + wanted.size());
            for (int i = 0; i < held.length; i++) {
                checkTrade(held[i], wanted.get(i), "trades[" + i + "]");
            }
        }

        if (expect.has("last")) {
            if (expect.get("last").isNull()) {
                assertNull(tape.last());
            } else {
                assertNotNull(tape.last(), "last() is null but the tape is not empty");
                checkTrade(tape.last(), expect.get("last"), "last()");
            }
        }

        // Last, because it empties the tape.
        if (expect.has("drain")) {
            assertEquals(expect.get("drain").get("count").asInt(), tape.drain().length, "drain");
            assertEquals(expect.get("drain").get("sizeAfter").asInt(), tape.size(), "size after drain");
            assertNull(tape.last());
            assertEquals(0, tape.drain().length);
        }
    }

    private void checkTrade(Trade trade, JsonNode expected, String where) {
        expected.propertyNames().forEach(key -> {
            JsonNode want = expected.get(key);
            switch (key) {
                case "price" -> assertEquals(want.asLong(), trade.price(), where + ".price");
                case "units" -> assertEquals(want.asLong(), trade.units(), where + ".units");
                case "restingId" -> assertEquals(want.asLong(), trade.resting().id(), where + ".restingId");
                case "aggressorId" -> assertEquals(want.asLong(), trade.aggressor().id(), where + ".aggressorId");
                case "restingOwnerId" ->
                        assertEquals(want.asLong(), trade.resting().ownerId(), where + ".restingOwnerId");
                case "aggressorOwnerId" ->
                        assertEquals(want.asLong(), trade.aggressor().ownerId(), where + ".aggressorOwnerId");
                case "at" -> {
                    assertNotNull(trade.at(), where + ".at: expected an instant, got null");
                    assertEquals(want.get("epochMilli").asLong(), trade.at().toEpochMilli(),
                            where + ".at is the wrong side's stamp — the trade happened when "
                            + "the aggressor arrived, not when the quote it took was posted");
                }
                default -> fail(where + ": fixture asks for unknown key " + key);
            }
        });
    }

    @Test
    void thereAreFixturesToRun() throws Exception {
        // Guard the guard: a bad path would report everything passing.
        assertTrue(fixtures().count() >= 4, "too few behaviour fixtures found in " + FIXTURES);
    }

    @Test
    void everyFixtureTypeHasAnAggregator() throws Exception {
        List<String> unmapped = fixtures()
                .map(f -> f.document().get("type").asString())
                .distinct()
                .filter(type -> !AGGREGATORS.contains(type))
                .toList();
        assertTrue(unmapped.isEmpty(), "fixtures exist for " + unmapped + ", which nothing here drives");
    }
}

File: sdks/typescript/test/behaviour-fixtures.test.ts
/**
 * Every behaviour fixture, driven through this SDK's aggregators.
 *
 * The wire fixtures next door compare *parsed field values*: one payload in,
 * one set of fields out. They say nothing about what Book and Tape do
 * with a sequence of them, which is where the three SDKs have actually been
 * wrong together — a book that double-counts a cancel and a tape that holds its
 * trades backwards both parse every field correctly.
 *
 * So these are inputs and answers rather than payloads and fields: a market, a
 * list of update steps, and what the aggregator must hold at the end. Java,
 * Python and TypeScript each run all of them, so a behaviour cannot be right in
 * one SDK and wrong in another without saying so.
 *
 * See `sdks/fixtures/README.md`.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import { readdirSync, readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

import { parseOrder } from "../src/client.js";
import { Book } from "../src/orderbook.js";
import { Tape, type Trade } from "../src/trades.js";
import type { Market, Order } from "../src/types.js";

const DIR = fileURLToPath(new URL("../../fixtures/behaviour/", import.meta.url));
const FILES = readdirSync(DIR).filter((f) => f.endsWith(".json")).sort();

interface Step { orders: Record<string, unknown>[]; clear?: boolean; adds?: number; note?: string }
interface Fixture {
  type: string;
  market: { id: number; symbol: string };
  deliveredIds: number[];
  steps: Step[];
  expect: Record<string, unknown>;
}

const AGGREGATOR_TYPES = new Set(["Book", "Tape"]);

function market(doc: Fixture): Market {
  return { id: doc.market.id, symbol: doc.market.symbol } as Market;
}

function drive(doc: Fixture): Book | Tape {
  const aggregator = doc.type === "Book"
    ? new Book(market(doc))
    : new Tape(market(doc));

  doc.steps.forEach((step, index) => {
    if (step.clear) aggregator.clear();

    const orders = step.orders.map(parseOrder) as Order[];

    // A step that declares `refused` must raise rather than apply: an order that names no side cannot be placed on a book, and guessing one files it under the offer side silently. The step after it asserts the book was left alone.
    if (step.refused !== undefined) {
      assert.throws(() => aggregator.update(orders), (e: Error) =>
        e.message.includes(step.refused as string),
        `step ${index} (${step.note ?? ""}) must be refused, not guessed at`);
      return;
    }

    const added = aggregator.update(orders);

    // What update() reports it added is what Desk dispatches onTrade
    // from. A step that declares `adds` pins it — including the zero, which is
    // the update a handler must stay silent through.
    if (step.adds !== undefined) {
      const count = added instanceof Map
        ? [...added.values()].reduce((n, t) => n + t.length, 0)
        : (added as unknown[]).length;
      assert.equal(count, step.adds,
        `step ${index} (${step.note ?? ""}) reported ${count} new trades, ` +
          `expected ${step.adds}`);
    }
  });
  return aggregator;
}

function epochMilli(value: Date | null, where: string): number {
  assert.notEqual(value, null, `${where}: expected an instant, got null`);
  return value!.getTime();
}

function checkTrade(actual: Trade, expected: Record<string, any>, where: string): void {
  const readers: Record<string, (t: Trade) => unknown> = {
    price: (t) => t.price,
    units: (t) => t.units,
    restingId: (t) => t.resting.id,
    aggressorId: (t) => t.aggressor.id,
    restingOwnerId: (t) => t.resting.ownerId,
    aggressorOwnerId: (t) => t.aggressor.ownerId,
  };
  for (const [key, want] of Object.entries(expected)) {
    if (key === "at") {
      assert.equal(
        epochMilli(actual.at, `${where}.at`), (want as any).epochMilli,
        `${where}.at is the wrong side's stamp — the trade happened when the ` +
          `aggressor arrived, not when the quote it took was posted`,
      );
      continue;
    }
    assert.ok(readers[key], `${where}: fixture asks for unknown key ${key}`);
    assert.deepEqual(readers[key]!(actual), want, `${where}.${key}`);
  }
}

function checkOrderBook(book: Book, expect: Record<string, any>): void {
  const readers: Record<string, () => unknown> = {
    bestBuyPrice: () => book.bestBuyPrice(),
    bestBuyUnits: () => book.bestBuyUnits(),
    bestSellPrice: () => book.bestSellPrice(),
    bestSellUnits: () => book.bestSellUnits(),
    hasValueBuy: () => book.hasValue("BUY"),
    hasValueSell: () => book.hasValue("SELL"),
    buyLevels: () => book.buyLevels().map(([p, u]) => [p, u]),
    sellLevels: () => book.sellLevels().map(([p, u]) => [p, u]),
  };
  for (const [key, want] of Object.entries(expect)) {
    assert.ok(readers[key], `fixture asks for unknown key ${key}`);
    assert.deepEqual(readers[key]!(), want, key);
  }
}

function checkTrades(tape: Tape, expect: Record<string, any>): void {
  const held = tape.mostRecentTrades();

  if ("size" in expect) assert.equal(tape.size(), expect["size"]);

  if ("trades" in expect) {
    assert.equal(held.length, expect["trades"].length,
      `tape holds ${held.length} trades, expected ${expect["trades"].length}`);
    held.forEach((actual, i) => checkTrade(actual, expect["trades"][i], `trades[${i}]`));
  }

  if ("last" in expect) {
    if (expect["last"] === null) {
      assert.equal(tape.last(), null);
    } else {
      assert.notEqual(tape.last(), null, "last() is null but the tape is not empty");
      checkTrade(tape.last()!, expect["last"], "last()");
    }
  }

  // Last, because it empties the tape.
  if ("drain" in expect) {
    assert.equal(tape.drain().length, expect["drain"].count);
    assert.equal(tape.size(), expect["drain"].sizeAfter);
    assert.equal(tape.last(), null);
    assert.deepEqual(tape.drain(), []);
  }
}

for (const file of FILES) {
  const doc: Fixture = JSON.parse(readFileSync(DIR + file, "utf8"));

  test(`behaviour: ${file.replace(/\.json$/, "")}`, () => {
    const delivered = doc.steps.flatMap((s) => s.orders.map((o) => o["id"]));
    assert.deepEqual(
      delivered, doc.deliveredIds,
      "the fixture's input is not in the order it declares. deliveredIds is " +
        "what stops a case being quietly reordered into one that proves nothing " +
        "— the trades-ordering fixture is only a test at all because its input " +
        "arrives newest first.",
    );

    const aggregator = drive(doc);
    if (doc.type === "Book") checkOrderBook(aggregator as Book, doc.expect);
    else checkTrades(aggregator as Tape, doc.expect);
  });
}

test("there are behaviour fixtures to run", () => {
  assert.ok(FILES.length >= 4, `only found ${FILES.length} behaviour fixtures`);
});

test("every behaviour fixture type has an aggregator", () => {
  const types = new Set(FILES.map((f) => JSON.parse(readFileSync(DIR + f, "utf8")).type));
  const unmapped = [...types].filter((t) => !AGGREGATOR_TYPES.has(t));
  assert.deepEqual(unmapped, [], `fixtures exist for ${unmapped}, which nothing here drives`);
});
